### PR TITLE
[Inference]Support sentence transformers clip

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

Follow up for #408, support the inference of sentence transformers clip.

**[Compilation]** 

* With CLI

```bash
optimum-cli export neuron -m sentence-transformers/clip-ViT-B-32 --sequence_length 64 --text_batch_size 3 --image_batch_size 1 --num_channels 3 --height 224 --width 224 --task feature-extraction --library-name sentence_transformers --subfolder 0_CLIPModel clip_emb/
```

* Code

```python
from optimum.neuron import NeuronModelForSentenceTransformers

# [Compile]
model_id = "sentence-transformers/clip-ViT-B-32"

# configs for compiling model
input_shapes = {
    "num_channels": 3,
    "height": 224,
    "width": 224,
    "text_batch_size": 3,
    "image_batch_size": 1,
    "sequence_length": 64,
}

emb_model = NeuronModelForSentenceTransformers.from_pretrained(
    model_id, subfolder="0_CLIPModel", export=True, library_name="sentence_transformers", dynamic_batch_size=False, **input_shapes
)

# Save locally or upload to the HuggingFace Hub
save_directory = "clip_emb/"
emb_model.save_pretrained(save_directory)
```

**[Inference]** 

```python
from PIL import Image
from sentence_transformers import util
from transformers import CLIPProcessor

from optimum.neuron import NeuronModelForSentenceTransformers

save_directory = "clip_emb"
emb_model = NeuronModelForSentenceTransformers.from_pretrained(save_directory)

processor = CLIPProcessor.from_pretrained(save_directory)
inputs = processor(
    text=["Two dogs in the snow", 'A cat on a table', 'A picture of London at night'], images=Image.open("two_dogs_in_snow.jpg"), return_tensors="pt", padding=True
)  

outputs = emb_model(**inputs)


# Compute cosine similarities
cos_scores = util.cos_sim(outputs.image_embeds, outputs.text_embeds)
print(cos_scores)

# tensor([[0.3072, 0.1016, 0.1095]])
```

**Caveat**

Since compiled models with dynamic batch size enabled only accept tensors with the same batch size, we cannot set `dynamic_batch_size=True` if the input texts and images have different batch sizes. And as `NeuronModelForSentenceTransformers` pad the inputs to the batch size used during the compilation, you could use a relatively larger batch_size during the compilation for flexibility with the trade-off of compute).

eg. if you want to encode 3 or 4 or 5 texts and 1 image,  you could set `text_batch_size = 5 = max(3, 4, 5)` and `image_batch_size = 1` during the compilation.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: docs/source/tutorials/sentence_transformers.mdx [hunk 1]
```diff
@@ -79,6 +79,83 @@ print(f"token embeddings: {token_embeddings.shape}") # torch.Size([1, 7, 384])
...
 
 For deploying these models in a production environment, refer to the [Amazon SageMaker Blog](https://www.philschmid.de/inferentia2-embeddings).
+
+
+## CLIP
+
+
+### Compile CLIP for AWS Inferentia2 
+
+You can compile CLIP models with Optimum Neuron either by using the `optimum-cli` or `NeuronModelForSentenceTransformers` class. Adopt one approache that you prefer:
```
Comment (michaelbenayoun) [inline docs/source/tutorials/sentence_transformers.mdx:92]:
```suggestion
You can compile CLIP models with Optimum Neuron either by using the `optimum-cli` or `NeuronModelForSentenceTransformers` class. Adopt one approach that you prefer:
```

File: docs/source/tutorials/sentence_transformers.mdx [hunk 2]
```diff
@@ -79,6 +79,83 @@ print(f"token embeddings: {token_embeddings.shape}") # torch.Size([1, 7, 384])
...
+
+* With CLI
```
Comment (michaelbenayoun) [inline docs/source/tutorials/sentence_transformers.mdx:94]:
```suggestion
* With the Optimum CLI
```

File: docs/source/tutorials/sentence_transformers.mdx [hunk 3]
```diff
@@ -79,6 +79,83 @@ print(f"token embeddings: {token_embeddings.shape}") # torch.Size([1, 7, 384])
...
+
+```bash
+optimum-cli export neuron -m sentence-transformers/clip-ViT-B-32 --sequence_length 64 --text_batch_size 3 --image_batch_size 1 --num_channels 3 --height 224 --width 224 --task feature-extraction --library-name sentence_transformers --subfolder 0_CLIPModel clip_emb/
+```
+
+* With `NeuronModelForSentenceTransformers` class
```
Comment (michaelbenayoun) [inline docs/source/tutorials/sentence_transformers.mdx:100]:
```suggestion
* With the `NeuronModelForSentenceTransformers` class
```

File: optimum/commands/export/neuronx.py [hunk 4]
```diff
@@ -145,6 +145,16 @@ def parse_args_neuronx(parser: "ArgumentParser"):
         type=int,
         help=f"Batch size {doc_input}",
     )
+    input_group.add_argument(
+        "--text_batch_size",
+        type=int,
+        help=f"Batch size of text inputs {doc_input} (Only applied for multi-modal models)",
```
Comment (michaelbenayoun) [inline optimum/commands/export/neuronx.py:151]:
```suggestion
        help=f"Batch size of the text inputs {doc_input} (Only applied for multi-modal models)",
```

File: optimum/commands/export/neuronx.py [hunk 5]
```diff
@@ -145,6 +145,16 @@ def parse_args_neuronx(parser: "ArgumentParser"):
...
+    )
+    input_group.add_argument(
+        "--image_batch_size",
+        type=int,
+        help=f"Batch size of vision inputs {doc_input} (Only applied for multi-modal models)",
```
Comment (michaelbenayoun) [inline optimum/commands/export/neuronx.py:156]:
```suggestion
        help=f"Batch size of the vision inputs {doc_input} (Only applied for multi-modal models)",
```

File: optimum/exporters/neuron/convert.py [hunk 6]
```diff
@@ -167,7 +167,7 @@ def validate_model_outputs(
     input_shapes = {}
     for axis in config.mandatory_axes:
         input_shapes[axis] = getattr(config, axis)
-    if config.dynamic_batch_size is True:
+    if config.dynamic_batch_size is True and "batch_size" in input_shapes:
```
Comment (michaelbenayoun) [inline optimum/exporters/neuron/convert.py:170]:
```suggestion
    if config.dynamic_batch_size and "batch_size" in input_shapes:
```

File: optimum/neuron/modeling.py [hunk 7]
```diff
@@ -218,28 +218,41 @@ def forward(
...
-                0
-            ]  # Remove padding on batch_size(0), and sequence_length(1)
-            # sentence_embedding -> (batch_size, hidden_size)
-            sentence_embedding = self.remove_padding([outputs[1]], dims=[0], indices=[input_ids.shape[0]])[
-                0
-            ]  # Remove padding on batch_size(0)
+            if "clip" in model_type:
+                text_embeds = self.remove_padding([outputs[0]], dims=[0], indices=[input_ids.shape[0]])[
+                    0
+                ]  # Remove padding on batch_size(0)
```
Comment (michaelbenayoun) [inline optimum/neuron/modeling.py:238]: What's `batch_size(0)`?

File: optimum/neuron/utils/misc.py [hunk 8]
```diff
@@ -548,3 +549,60 @@ def check_if_weights_replacable(
         raise RuntimeError(
             "Unable to replace weights of the neuron model since its weights and neff are not separated, please set `inline_weights_to_neff=Talse` when converting the model to Neuron format."
         )
+
+
+# Copied and adapted from https://github.com/huggingface/optimum/blob/d03ab100206cb9f0e62167a36ee6997424bb9bb5/optimum/utils/save_utils.py#L27
+# To remove once we can bump to a transformers release including https://github.com/huggingface/transformers/pull/29169
+def maybe_load_preprocessors(
```
Comment (michaelbenayoun) [inline optimum/neuron/utils/misc.py:556]: Maybe put a check of the transformers version, if it's above the coming release, then fail. This way we can actually catch that and remove it.

File: optimum/neuron/modeling.py (hunk 7 above)
Comment (JingyaHuang) [reply optimum/neuron/modeling.py:238]: I put the index of each dim in brackets to remind me which axes to pad or remove pad. For bs is 0, sequence length is usually 1, it's more helpful for other tensors like those for stable diffusion models with multiple axes.

File: optimum/neuron/utils/misc.py (hunk 8 above)
Comment (JingyaHuang) [reply optimum/neuron/utils/misc.py:556]: Sounds good!

Comment (JingyaHuang) [reply optimum/neuron/utils/misc.py:556]: Just checked the PR I submited to transformers, it's not merged yet and I'm not 100% percent sure that it should be accepted by trfrs maintainers, thus completely don't know what trfr version to put here. Let's keep it this way, I will keep my eyes on how it goes...

Comment (JingyaHuang) [reply optimum/neuron/utils/misc.py:556]: https://github.com/huggingface/transformers/pull/29169

**Review by michaelbenayoun — APPROVED** [review]
LGTM

**Pull request merged**